backend/regalloc/regalloc_split_state.ml

```diff
@@ -43,6 +43,176 @@ let log_renaming_info : indent:int -> t -> unit =
+      Cfg_with_infos.t ->
+      definitions_at_beginning:definitions_at_beginning ->
+      definitions_at_beginning * Label.t Stack.t =
+   fun cfg_with_infos ~definitions_at_beginning ->
+    if split_debug
+    then log ~indent:1 "MoveSpillsAndReloads.move_definitions_at_beginning_down";
+    let definitions_at_beginning = ref definitions_at_beginning in
+    let doms = Cfg_with_infos.dominators cfg_with_infos in
+    let stack = Stack.create () in
+    Cfg_dominators.iter_breadth_dominator_tree
+      doms.Cfg_dominators.dominator_tree ~f:(fun (label : Label.t) ->
+        Stack.push label stack;
+        let block = Cfg_with_infos.get_block_exn cfg_with_infos label in
+        match Label.Map.find_opt label !definitions_at_beginning with
+        | None -> ()
+        | Some (definitions : Reg.Set.t) -> (
+          if Option.is_none (destruction_point_at_end block)
+             && Option.is_none block.exn
+          then
+            let successor_labels =
+              Cfg.successor_labels ~normal:true ~exn:false block
+            in
+            match Label.Set.cardinal successor_labels with
+            | 1 ->
+              let successor_label = Label.Set.choose successor_labels in
+              let successor_block =
+                Cfg.get_block_exn
+                  (Cfg_with_infos.cfg cfg_with_infos)
+                  successor_label
+              in
+              if (not (Label.equal label successor_label))
+                 && Label.Set.cardinal successor_block.predecessors = 1
+              then
+                let to_move : Reg.Set.t =
+                  Reg.Set.filter
+                    (fun (reg : Reg.t) ->
+                      let occurs = occurs_block block reg in
+                      not occurs)
+                    definitions
+                in
+                if not (Reg.Set.is_empty to_move)
+                then (
+                  if split_debug
+                  then
+                    log ~indent:2 "moving %a from block %d to block %d"
+                      Printmach.regset to_move label successor_label;
+                  definitions_at_beginning
+                    := Label.Map.update label
+                         (function
+                           | None -> assert false
+                           | Some set -> Some (Reg.Set.diff set to_move))
+                         !definitions_at_beginning;
+                  definitions_at_beginning
+                    := Label.Map.update successor_label
+                         (function
+                           | None -> Some to_move
+                           | Some set -> Some (Reg.Set.union set to_move))
+                         !definitions_at_beginning)
+            | _ -> ()));
+    !definitions_at_beginning, stack
+
+  (** A destruction at the end of a block can be move up if:
+
+      - the block is not a trap handler;
+      - the block has only one predecessor;
+      - the predecessor has only one (normal) successor;
```

Comment:
```suggestion
      - the predecessor has only one (normal) successor and no exceptional successor;
```

Reply:
(Applied manually - because of `ocamlformat`.)